File: python/leetcode/025.py
class Solution(object):
    def reverseKGroup(self, head, k):
        if not head or k == 1:
            return head
        dummy = ListNode(0)
        dummy.next, ptr, n = head, dummy, 0
        while ptr.next:
            n += 1
            ptr = ptr.next
        pre = dummy
        while n >= k:
            ptr = pre.next
            for i in range(1, k):
                tmp = ptr.next
                ptr.next = tmp.next
                tmp.next = pre.next
                pre.next = tmp
            pre = ptr
            n -= k
        return dummy.next

File: cpp/leetcode/025.cpp
class Solution {
  public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        if (!head || k == 0) {
            return head;
        }
        int n = 0;
        for (ListNode *ptr = head; ptr; ptr = ptr->next) {
            ++n;
        }
        ListNode *dummy = new ListNode(0);
        dummy->next = head;
        ptr = dummy;
        while (n >= k) {
            ListNode *tail = ptr->next;
            for (int i = 0; i < k - 1; i++) {
                ListNode *tmp = tail->next;
                tail->next = tmp->next;
                tmp->next = ptr->next;
                ptr->next = tmp;
            }
            ptr = tail;
            n -= k;
        }
        return dummy->next;
    }
};